pkg/cache/sql/informer/indexer.go

```diff
@@ -0,0 +1,255 @@
+// AfterUpsert updates indices of an object
+func (i *Indexer) AfterUpsert(key string, obj any, tx db.TXClient) error {
+	// delete all
+	err := tx.StmtExec(tx.Stmt(i.deleteIndicesStmt), key)
+	if err != nil {
+		return err
+	}
+
+	// re-insert all
+	i.indexersLock.RLock()
+	defer i.indexersLock.RUnlock()
+	for indexName, indexFunc := range i.indexers {
+		values, err := indexFunc(obj)
+		if err != nil {
+			return err
+		}
+
+		for _, value := range values {
+			err = tx.StmtExec(tx.Stmt(i.addIndexStmt), indexName, value, key)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// add afterDelete
+/* Satisfy cache.Indexer */
+
+// Index returns a list of items that match the given object on the index function
+func (i *Indexer) Index(indexName string, obj any) ([]any, error) {
+	i.indexersLock.RLock()
+	defer i.indexersLock.RUnlock()
+	indexFunc := i.indexers[indexName]
+	if indexFunc == nil {
+		return nil, fmt.Errorf("index with name %s does not exist", indexName)
+	}
+
+	values, err := indexFunc(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(values) == 0 {
+		return nil, nil
+	}
+
+	// typical case
+	if len(values) == 1 {
+		return i.ByIndex(indexName, values[0])
+	}
+
+	// atypical case - more than one value to lookup
+	// HACK: sql.Statement.Query does not allow to pass slices in as of go 1.19 - create an ad-hoc statement
+	query := fmt.Sprintf(fmt.Sprintf(selectQueryFmt, i.GetName(), strings.Repeat(", ?", len(values)-1)))
+	stmt := i.Prepare(query)
+	defer i.CloseStmt(stmt)
+	// HACK: Query will accept []any but not []string
+	params := []any{indexName}
+	for _, value := range values {
+		params = append(params, value)
+	}
+
+	rows, err := i.QueryForRows(context.TODO(), stmt, params...)
+	if err != nil {
+		return nil, err
+	}
+	return i.ReadObjects(rows, i.GetType(), i.GetShouldEncrypt())
+}
+
+// ByIndex returns the stored objects whose set of indexed values
+// for the named index includes the given indexed value
+func (i *Indexer) ByIndex(indexName, indexedValue string) ([]any, error) {
+	rows, err := i.QueryForRows(context.TODO(), i.listByIndexStmt, indexName, indexedValue)
+	if err != nil {
+		return nil, err
+	}
+	return i.ReadObjects(rows, i.GetType(), i.GetShouldEncrypt())
+}
+
+// IndexKeys returns a list of the Store keys of the objects whose indexed values in the given index include the given indexed value
+func (i *Indexer) IndexKeys(indexName, indexedValue string) ([]string, error) {
+	i.indexersLock.RLock()
+	defer i.indexersLock.Unlock()
+	indexFunc := i.indexers[indexName]
+	if indexFunc == nil {
+		return nil, fmt.Errorf("Index with name %s does not exist", indexName)
+	}
+
+	rows, err := i.QueryForRows(context.TODO(), i.listKeysByIndexStmt, false, indexName, indexedValue)
+	if err != nil {
+		return nil, err
+	}
+	return i.ReadStrings(rows)
+}
+
+// ListIndexFuncValues wraps safeListIndexFuncValues and panics in case of I/O errors
+func (i *Indexer) ListIndexFuncValues(name string) []string {
```

Comment:
Trying to use the ListIndexFuncValues method here: https://github.com/tomleb/lasso-controller-notes/blob/f4b0ad310f76999edfadaf03c82f5308f6cdf3b6/simple-vai/main.go#L87-L101 and it returns an empty list of keys instead of a list of all configmap keys.

I have tried using an upstream `cache.SharedIndexInformer` and the same example works fine. Am I missing something in the example that would make it work?

Reply:
These were all due to bugs which I have now found and resolved. Should add integration tests for this. It may be difficult to include in a rancher integration test but they can be setup for lasso. Here are the reasons for each failure:
1.) When [adding encryption](https://github.com/rmweir/lasso/blob/vai/pkg/cache/sql/db/client.go#L219) the object scan was adjusted to accept the object columns and the 3 encryption related columns. The [select query](https://github.com/rmweir/lasso/blob/vai/pkg/cache/sql/informer/indexer.go#L18) was not updated to also select these columns so it ended up passing a number of rawbytes pointers that didn't match the number of values queried.
2.) Same as 1, but for the [list by index query](https://github.com/rmweir/lasso/blob/vai/pkg/cache/sql/informer/indexer.go#L35).
3.) Used Unlock() when RUnlock() is needed.
4.) For some reason [false was being passed as a param to the query](https://github.com/rmweir/lasso/blob/vai/pkg/cache/sql/informer/indexer.go#L211).
5.) [Same as 4 when querying for index values](https://github.com/rmweir/lasso/blob/vai/pkg/cache/sql/informer/indexer.go#L229).